### DISCUSSION TITLE
RDKit logs everything in standardize function when using python multithreading

### DISCUSSION THREAD

**Discussion by ewbell94** [rdkit/rdkit#6771] [Q&A]
I've been using RDKit to preprocess and parameterize a set of small molecules for an ML application.  Since the SDFs are pretty huge, I've been trying to multithread the parameterization part, and in a lazy way, since the parameterized molecules are much less memory intensive when compared to Mol objects.  The following is a simplified version of my code:
```
import multiprocessing as mp
from rdkit import Chem
from rdkit.Chem.PropertyMol import PropertyMol
from rdkit.Chem.MolStandardize import rdMolStandardize as rds

def standardize(mol):
    try:
        uncharger = rds.Uncharger()
        mol = rds.Cleanup(mol)
        mol = rds.FragmentParent(mol)
        mol = uncharger.uncharge(mol)
        return mol
    except Exception as e:
        print(str(e))
        return None
        
def readMol(sdffile):
    suppl = Chem.SDMolSupplier(sdffile)
    for mol in suppl:
        if mol != None:
            mol = standardize(mol)
            if mol != None:
                yield PropertyMol(mol) #otherwise properties will get eaten by pickle

if __name__ == "__main__":
    paramMol = []
    def addRes(result):
        if result != None:
            paramMol.append(result)

    p = mp.pool.ThreadPool()
    for mol in readMol("mols.sdf"):
        #parameterize(mol) is not provided but converts the Mol object into a pytorch tensor
        p.apply_async(parameterize, (mol,), callback=addRes)
    p.join()
    p.close()
```
My problem is that when p is a `mp.pool.ThreadPool()`, RDKit logs a bunch of Info logs to stdout, particularly detailing the steps it took in the `standardize(mol)` function.  I have tried calling various `rdkit.rdBase` functions such as `DisableLog()` and `BlockLogs()` both in the main and the `standardize(mol)` function to no effect.  I have even had it print `LogStatus()` immediately prior to calling the standardization functions and tell me that all logs are disabled and then subsequently logging that the steps of the `standardize` function were occurring.  The final baffling piece to this puzzle is that if the `mp.pool.ThreadPool()` object is changed to a `mp.Pool()` object, the logs go away (but I want to use threads as there seems to be some latency on the callback in a multiprocessing scheme).  Is this intended behavior, or is there something thread-unsafe going on in my code that I'm not seeing?

**Comment by bp-kelley** [comment]
What version of RDKit are you using?  MolStandardize has gotten a lot quieter recently, I have an older one:

'2021.03.1'

That was quite noisy.  But the current one (2023.xx) is a lot quieter.

In general, you can block rdkit logs as follows:

```
from rdkit.rdBase import BlockLogs
def func():
      blocker = rdBase.BlockLogs()
```

You have to bind it to a variable because it is scoped.  When it goes out of scope, it returns the logs to their previous state.

Also the following comment isn't strictly true:

```
PropertyMol(mol) #otherwise properties will get eaten by pickle
```

Note: You shouldn't need to use PropertyMol with ThreadPools, they shouldn't be pickled at all.

When using pickle, you can add the following to preserve properties:

```
 Chem.SetDefaultPickleProperties(Chem.PropertyPickleOptions.MolProps)
```

Finally, many functions in the RDKit Python wrapper don't release the GIL, you may have better luck with multiprocessing.Pool.

**Reply by ewbell94** [reply]
Thank you for pointing out the pickle comment; that's a leftover from when I was doing multiprocessing instead of multithreading (in which the memory *is* pickled to the child process).  I wasn't aware of this pickle properties variable, and that's definitely a much easier solution than having to wrap everything in `PropertyMol` objects.

I'm using 2022.03.5, and have tried using `BlockLogs()` to turn off logs, both in the global and local scope of the function, yet it's still giving me logs. The

You're right about the fact that the GIL *should* be a problem in my multithreading, but for some inexplicable reason it isn't.  My best guess is that since the CPU bottleneck is mostly a UFF geometry optimization in the `parameterize` function, it's just chugging away on the C++ side instead of needing the python interpreter, but I really have no clue why multithreading is working better than multiprocessing.  Anecdotally I'm seeing 100% CPU usage per core when multithreading and ~20% CPU usage when multiprocessing.

**Comment by kienerj** [comment]
No an answer to the question but I don't think using multi threading here will help due to GIL and you will need to do multiprocessing.

Please feel free to either use my [experimental tool](https://github.com/kienerj/rdkit_parallel_tools) to make such tasks easier or look at [this blog post](https://baoilleach.blogspot.com/2020/05/python-patterns-for-processing-large.html) that inspired above tool.

Core feature is to read the sdf as strings of mol blocks  incl. properties and submit these as text to the individual processes. Therefore no need to pickle molecules at all. Exception handling will be a bit tricky but in general this will speed up things pretty much to as many cores as you have available.
